### PR TITLE
fix: adds exception check for debug_setHead

### PR DESCRIPTION
**Pull request by bitwise-constructs**
### What I did
adds exception check for debug_setHead when dealing with nodes that don't support it

fixes: #2215
Fixes: APE-1797

### How I did it

added a cached property to provider for skipping debug_setHead restore if not supported

### How to verify it

run on a forked gnosis chain

### Checklist

- [x] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: src/ape_node/provider.py [hunk 1]
```diff
@@ -405,7 +409,10 @@ def restore(self, snapshot_id: SnapshotID):
             logger.error("Unable to set head to future block.")
             return
 
-        self.make_request("debug_setHead", [block_number_hex_str])
+        try:
+            self.make_request("debug_setHead", [block_number_hex_str])
+        except Exception:
+            self._supports_debug_set_head = False
```
Comment (fubuloubu) [inline src/ape_node/provider.py:415]: Should this raise `NotImplementedError` instead at this point?

File: src/ape_node/provider.py [hunk 2]
```diff
@@ -405,7 +409,10 @@ def restore(self, snapshot_id: SnapshotID):
             logger.error("Unable to set head to future block.")
             return
 
-        self.make_request("debug_setHead", [block_number_hex_str])
+        try:
+            self.make_request("debug_setHead", [block_number_hex_str])
+        except Exception:
```
Comment (antazoey) [inline src/ape_node/provider.py:414]:
```suggestion
        except NotImplementedError:
```

File: src/ape_node/provider.py (hunk 1 above)
Comment (antazoey) [reply src/ape_node/provider.py:415]: that is actually the exception that was already be raised!

We could move the handling `chain.snapshot` though: https://github.com/ApeWorX/ape/blob/main/src/ape/managers/chain.py#L1607

File: src/ape_node/provider.py (hunk 2 above)
Comment (antazoey) [reply src/ape_node/provider.py:414]: Perhaps we could re-raise the exception and also raise during `not self._supports_debug_set_head:` and then catch `NotImplementedError` in `chain.snapshot` method instead

Comment (antazoey) [reply src/ape_node/provider.py:414]: or perhaps the handling should just happen in he pytest integration

**Review by antazoey — APPROVED** [review]

**Pull request merged**